### DISCUSSION TITLE
Does the `proxy_ip_parser` not work with IPv6 addresses properly?

### DISCUSSION THREAD

**Discussion by victor-sudakov** [roadrunner-server/roadrunner#1516] [Q&A] [answered]
Dear Colleagues,

I have the proxy_ip_parser enabled in roadrunner 2.12.3
```yaml
http:
  address: 0.0.0.0:8060
  middleware: ["http_metrics", "headers", "gzip", "proxy_ip_parser"]
```
I have noticed a problem: when the `X-Real-IP` header contains an IPv6 address, roadrunner logs this IPv6 address to its log correctly but does not set the `$_SERVER['REMOTE_ADDR']` variable for the PHP workers, they receive the variable as empty. When however the `X-Real-IP` header contains an IPv4 address, the `$_SERVER['REMOTE_ADDR']` variable as seen from the PHP workers has the expected value.

Is it a possible bug or some misconfiguration on my part?

A sample log entry with IPv6:

```
2023-03-27T03:14:53.913Z	INFO	http        	http log	{"status": 200, "method": "HEAD", "URI": "/YYY/XXXXXXXXXXXXX", "remote_address": "2a01:111:f400:fe5a::100", "read_bytes": 0, "write_bytes": 0, "start": "2023-03-27T03:14:53.432Z", "elapsed": "481.145642ms"}
```

**Comment by rustatian** [comment]
Hey @victor-sudakov 👋🏻 
I double checked the tests, it seems that the RR is passing the value correctly to the PHP. I'm not sure what is `$_SERVER`, since the RR passes data to the PHP worker via JSON. It may be that the PHP worker is setting this value incorrectly, you can verify this by checking the RR's payload in the PHP worker.

**Comment by vladimir-vv** [comment, marked as answer]
Hello @rustatian 

We use RR to launch Laravel app.
"laravel/framework", "version": "v9.52.4"
"spiral/roadrunner-laravel", "version": "v5.11.1"

I added error_log() into "vendor/spiral/roadrunner-http/src/HttpWorker.php"
in function waitRequest() to dump payload.

Result with IPv4:
```json
{
  "remoteAddr": "X.X.196.87",
  "protocol": "HTTP/1.1",
  "method": "GET",
  "uri": "http://SITE.com/",
  "headers": {
    "Cf-Connecting-Ip": [
      "X.X.196.87"
    ],
    "X-Scheme": [
      "https"
    ],
    "User-Agent": [
      "curl/7.87.0"
    ],
    "X-Request-Id": [
      "899100e44f1f6d7da07580165192b934"
    ],
    "X-Real-Ip": [
      "X.X.196.87"
    ],
    "X-Forwarded-For": [
      "X.X.196.87"
    ],
    "X-Forwarded-Scheme": [
      "https"
    ],
    "Cf-Visitor": [
      "{\"scheme\":\"https\"}"
    ],
    "X-Forwarded-Host": [
      "SITE.com"
    ],
    "X-Forwarded-Proto": [
      "https"
    ],
    "X-Original-Forwarded-For": [
      "X.X.196.87"
    ],
    "Accept": [
      "*/*"
    ],
    "Cf-Ipcountry": [
      "RU"
    ],
    "X-Forwarded-Port": [
      "443"
    ],
    "Accept-Encoding": [
      "gzip"
    ],
    "Cf-Ray": [
      "7aed6de6aba12de4-DME"
    ],
    "Cdn-Loop": [
      "cloudflare"
    ]
  },
  "cookies": {},
  "rawQuery": "",
  "parsed": false,
  "uploads": null,
  "attributes": {}
}
```

Result with IPv6:
```json
{
  "remoteAddr": "",
  "protocol": "HTTP/1.1",
  "method": "GET",
  "uri": "http://SITE.com/",
  "headers": {
    "Cf-Connecting-Ip": [
      "2001:19f0:5001:21dc::10"
    ],
    "X-Scheme": [
      "https"
    ],
    "User-Agent": [
      "curl/7.87.0"
    ],
    "X-Request-Id": [
      "2af82c002779440ff68431b5e9ac3fd7"
    ],
    "X-Real-Ip": [
      "2001:19f0:5001:21dc::10"
    ],
    "X-Forwarded-For": [
      "2001:19f0:5001:21dc::10"
    ],
    "X-Forwarded-Scheme": [
      "https"
    ],
    "Cf-Visitor": [
      "{\"scheme\":\"https\"}"
    ],
    "X-Forwarded-Host": [
      "SITE.com"
    ],
    "X-Forwarded-Proto": [
      "https"
    ],
    "X-Original-Forwarded-For": [
      "2001:19f0:5001:21dc::10"
    ],
    "Accept": [
      "*/*"
    ],
    "Cf-Ipcountry": [
      "NL"
    ],
    "X-Forwarded-Port": [
      "443"
    ],
    "Accept-Encoding": [
      "gzip"
    ],
    "Cf-Ray": [
      "7aed6717b818b926-AMS"
    ],
    "Cdn-Loop": [
      "cloudflare"
    ]
  },
  "cookies": {},
  "rawQuery": "",
  "parsed": false,
  "uploads": null,
  "attributes": {}
}
```

As you see "remoteAddr" field is empty in case of IPv6. 
If you need any additional information we are ready to provide it.

**Reply by rustatian** [reply]
I did a double check on the request, RR sends this data to the PHP worker:

```json
{"remoteAddr":"::1","protocol":"HTTP/1.1","method":"GET","uri":"http://[0:0:0:0:0:0:0:1]:10684/?hello=world","headers":{"User-Agent":["Go-http-client/1.1"],"Accept-Encoding":["gzip"]},"cookies":{},"rawQuery":"hello=world","parsed":false,"uploads":null,"attributes":{}}
```

**Reply by rustatian** [reply]
Is it possible that you have not added your IPv6 address to the list of trusted subnets?

**Reply by vladimir-vv** [reply]
We have this configuration "proxy_ip_parser" for now:

```yaml
http:
  address: 0.0.0.0:8060
  middleware: ["http_metrics", "headers", "gzip", "proxy_ip_parser"]
  trusted_subnets: [ "0.0.0.0/0", "::/0" ]
```

RR log:
```
2023-03-28T09:07:24.279Z	INFO	http        	http log	{"status": 200, "method": "GET", "URI": "/", "remote_address": "2001:19f0:5001:21dc::10", "read_bytes": 0, "write_bytes": 2607, "start": "2023-03-28T09:07:24.270Z", "elapsed": "8.534272ms"}
```

And "remoteAddr"  empty.

**Reply by rustatian** [reply]
Tested with your `trusted_subnets`:

```json
{"remoteAddr":"::1","protocol":"HTTP/1.1","method":"GET","uri":"http://[0:0:0:0:0:0:0:1]:10684/?hello=world","headers":{"User-Agent":["Go-http-client/1.1"],"Accept-Encoding":["gzip"]},"cookies":{},"rawQuery":"hello=world","parsed":false,"uploads":null,"attributes":{}}
```

Definitely a problem in the PHP package. By the way, officially we only support Spiral Framework.

**Reply by vladimir-vv** [reply]
Ok. I made simple investigation.
1) created new app: ``composer create-project spiral/app test-spiralfw``
2) enabled proxy_ip_parser and added code to show remoteAddr
Here is the diff:

```diff
diff -ru test-spiralfw0/app/src/Endpoint/Web/HomeController.php test-spiralfw/app/src/Endpoint/Web/HomeController.php
--- test-spiralfw0/app/src/Endpoint/Web/HomeController.php	2023-03-28 18:06:13.321358483 +0700
+++ test-spiralfw/app/src/Endpoint/Web/HomeController.php	2023-03-28 17:53:51.431590191 +0700
@@ -7,6 +7,7 @@
 use Exception;
 use Spiral\Prototype\Traits\PrototypeTrait;
 use Spiral\Router\Annotation\Route;
+use Spiral\Http\Request\InputManager;
 
 /**
  * Simple home page controller. It renders home page template and also provides
@@ -22,9 +23,9 @@
 
 
     #[Route(route: '/', name: 'index')]
-    public function index(): string
+    public function index(InputManager $input): string
     {
-        return $this->views->render('home');
+        return $this->views->render('home', ['ip' => $input->remoteAddress()]);
     }
 
     /**
diff -ru test-spiralfw0/app/views/home.dark.php test-spiralfw/app/views/home.dark.php
--- test-spiralfw0/app/views/home.dark.php	2023-03-28 18:05:58.571236107 +0700
+++ test-spiralfw/app/views/home.dark.php	2023-03-28 17:57:20.226814219 +0700
@@ -8,6 +8,8 @@
 <define:body>
     <div class="wrapper">
         <div class="placeholder">
+IP: [<?=$ip?>]
+
             <img src="/images/logo.svg" alt="Framework Logotype" width="200px"/>
             <h2>[[Welcome to Spiral Framework]]</h2>
 
diff -ru test-spiralfw0/.rr.yaml test-spiralfw/.rr.yaml
--- test-spiralfw0/.rr.yaml	2023-03-28 18:06:13.354692093 +0700
+++ test-spiralfw/.rr.yaml	2023-03-28 18:00:51.368650253 +0700
@@ -6,6 +6,8 @@
     middleware:
         - gzip
         - static
+        - proxy_ip_parser
+    trusted_subnets: [ "0.0.0.0/0", "::/0" ]
     static:
         dir: public
         forbid:
Only in test-spiralfw/runtime: logs
Only in test-spiralfw/runtime: snapshots
```

Test 1:
``curl http:///127.0.0.1:8080/``
IP: [127.0.0.1]
``2023-03-28T18:00:29.003+0700	INFO	http        	http log	{"status": 200, "method": "GET", "URI": "/", "remote_address": "127.0.0.1:55070", "read_bytes": 0, "write_bytes": 925, "start": "2023-03-28T18:00:29.003+0700", "elapsed": "733.134µs"}``

Test 2:
``curl -H 'X-Real-Ip: 2001:19f0:5001:21dc::10' http:///127.0.0.1:8080/``
IP: []
``2023-03-28T18:02:01.133+0700	INFO	http        	http log	{"status": 200, "method": "GET", "URI": "/", "remote_address": "2001:19f0:5001:21dc::10", "read_bytes": 0, "write_bytes": 916, "start": "2023-03-28T18:02:01.132+0700", "elapsed": "709.179µs"}``

Test 3
``curl -H 'Cf-Connecting-Ip: who let the dogs out' http:///127.0.0.1:8080/``
IP: [who let the dogs out]
``2023-03-28T18:21:04.353+0700	INFO	http        	http log	{"status": 200, "method": "GET", "URI": "/", "remote_address": "who let the dogs out", "read_bytes": 0, "write_bytes": 936, "start": "2023-03-28T18:21:04.352+0700", "elapsed": "1.0118ms"}``

Test 4
``curl -H 'Cf-Connecting-Ip: 2001:19f0:5001:21dc::10' http:///127.0.0.1:8080/``
IP: []
``2023-03-28T18:21:48.570+0700	INFO	http        	http log	{"status": 200, "method": "GET", "URI": "/", "remote_address": "2001:19f0:5001:21dc::10", "read_bytes": 0, "write_bytes": 916, "start": "2023-03-28T18:21:48.569+0700", "elapsed": "940.981µs"}``

Our app is behind cloudflare and look like there is a problem in getting right ip address. 
Or may be we have incorrect settings.  🤔

**Reply by rustatian** [reply]
The problem is that you are specifying an incorrect IPv6 -> `2001:19f0:5001:21dc::10`.
This IP address is correct to display as TEXT. But `remote_address` is URI, which forces you to use square brackets, like this `[2001:19f0:5001:21dc]:10`.

Standards: 
- https://www.rfc-editor.org/rfc/rfc5952#section-6
From docs: This is due to the "::"usage in IPv6 addresses. This style is NOT RECOMMENDED because of its ambiguity.  The [] style as expressed in [[RFC3986](https://www.rfc-editor.org/rfc/rfc3986)] SHOULD be employed, and is the default unless otherwise specified.

- https://www.rfc-editor.org/rfc/rfc3986#section-3.2.2

**Reply by rustatian** [reply]
Here is the correct `curl` request: `curl -H 'Cf-Connecting-Ip: [2001:19f0:5001:21dc]:10' http:///127.0.0.1:8080/`

**Reply by vladimir-vv** [reply]
The "10" is not a port number. It's a part of a real-life IPv6 address ("2001:19f0:5001:21dc::10") of a host we are testing from. You can ping it. There is no port number in it.
The 'Cf-Connecting-Ip' is not our custom header. This header is added by Cloudflare. Here you can find documentation about their headers: [Cloudflare Fundamentals: HTTP request headers](https://developers.cloudflare.com/fundamentals/get-started/reference/http-request-headers/)
We do not control Cloudflare headers.

Another example with real-life IPv6:
`curl -H 'X-Real-Ip: 2001:470:ecba:3:2c61:5e11:910d:772d' http:///127.0.0.1:8080/`
And RemoteAddr is also empty.

**Reply by vladimir-vv** [reply]
You wrote 
> But remote_address is URI

May be here is the problem. That you suppose that remote_address is URI.

From RFC you provided:

> [3](https://www.rfc-editor.org/rfc/rfc3986#section-3).  Syntax Components
>    The generic URI syntax consists of a hierarchical sequence of
>    components referred to as the scheme, authority, path, query, and
>    fragment.
>       URI         = scheme ":" hier-part [ "?" query ] [ "#" fragment ]

Remote address - should be just an IPv4 or IPv6 address.

**Reply by rustatian** [reply]
> The "10" is not a port number. 

I did not say that `10` was a port number. I just sent you a corrected example.

> Another example with real-life IPv6...

I'm not saying your example isn't real, I'm saying where is the problem. The problem is in the form of the IPv6 address.

> Remote address - should be just an IPv4 or IPv6 address.

There is no term `just`. It can be TEXT or URI. We parse it like a URI and that's why you don't get the IP address in the `remoteAdderss` field. Because the IP address you sent is not a proper IPv6 URI address, because you missed **[** and **]** in it. For example: `[0:0:0:0:0:0:0:1]:10684` or `[::1]:59964` are the correct IPv6 URIs according to the standard (address+port).

From docs: A host identified by an Internet Protocol literal address, version 6
[[RFC3513](https://www.rfc-editor.org/rfc/rfc3513)] or later, is distinguished by enclosing the IP literal **within square brackets ("[" and "]")**.

**Reply by vladimir-vv** [reply]
Well, Cloudflare and any other company will not provide brackets in such headers.
And the bug (after tests we made looks like this is a bug) will still exists.
Regardless that you think that remoteAddress is URI.

We used real-life examples of headers.

**Reply by rustatian** [reply]
Remote address **is** URI (I'm not guessing here). But you're missing the point, that the value in the HEADER is not an URI (if it's not like `X-Forwarded-URI`). But still, if you think, that something here is incorrect PR's are always welcome with the documentation.
Here you are the xff middleware: https://github.com/roadrunner-server/proxy_ip_parser
And http plugin: https://github.com/roadrunner-server/http

**Reply by victor-sudakov** [reply]
> Remote address is URI

Unfortunately CloudFlare does not share this idea and inserts its headers in the following format, without any brackets:
`Cf-Connecting-Ip: 2001:470:ecba:3:d7a3:c01a:c3c7:b9b0`

Even if they are in error, I see no way to change that. And with this de-facto practice, this middleware is useless for IPv6.

**Reply by vladimir-vv** [reply]
I'm not talking about Golang types (about RemoteAddr). 
I'm talking about common headers, nginx and php variables.
Sorry, I already don't understand do you want the RR to work properly or not.

PHP docs:
'REMOTE_ADDR' - The IP address from which the user is viewing the current page.
'REMOTE_PORT' - The port being used on the user's machine to communicate with the web server. 

Nginx docs:
$remote_addr - client address 
$remote_port - client port

**Reply by vladimir-vv** [reply]
I can suppose that FetchIP() in rr-plugins could be a little bit different. For example:

```golang
func FetchIP(pair string) string {
	if !strings.ContainsRune(pair, ':') {
		return pair
	}
	addr, _, err := net.SplitHostPort(pair)
	if err == nil {
		return addr
	}
	ip := net.ParseIP(pair)
	if ip == nil {
		return ""  // or other error processing in such case
	}
	return ip.String()
}
...
	fmt.Println(FetchIP("127.0.0.1"))
	fmt.Println(FetchIP("127.0.0.1:8080"))
	fmt.Println(FetchIP("::1"))
	fmt.Println(FetchIP("2001:19f0:5001:21dc::10"))
	fmt.Println(FetchIP("[2001:19f0:5001:21dc::10]:8080"))

```
Output:
```
127.0.0.1
127.0.0.1
::1
2001:19f0:5001:21dc::10
2001:19f0:5001:21dc::10
```

But of course it's up to you to decide how to get IP(v4/v6) for remote_addr to provide to PHP.

**Reply by rustatian** [reply]
> I already don't understand do you want the RR to work properly or not.

1. I pointed you both to the actual standard, spend time figuring out where the problem is. Please respect others in this discussion.

2. I pointed out that the value passed via HEADERS cannot be parsed as a valid IPv6 address and assigned to the remote address, because the value passed without brackets and remote address is required to fit in this standard because it is parsed with the `net` package.

3. I also pointed out that to change something, you should provide the **documentation** about that (PRs are also welcome), but not your personal opinion, like `do you want the RR to work properly or not`.

**Reply by victor-sudakov** [reply]
@rustatian 
> because the value passed without brackets and remote address is required to fit in this standard because it is parsed with the net package.

Why do you think CloudFlare is passing this address without brackets? Are they also mistaken?

**Reply by vladimir-vv** [reply]
@rustatian, I respect your work. I'm doing my best to localize the problem and safe your time.
I don't feel like I'm good enough to suggest pull requests because in my opinion it should be well tested. 
It is obvious that you are better know how RR works.
Facts we have:
1) headers that are common use don't use brackets for IPv6 to provide IP address
2) php, as a child of RR process expects that remote_addr should be IP address, no brackets

**Reply by rustatian** [reply]
> @rustatian, I respect your work. I'm doing my best to localize the problem and safe your time. I don't feel like I'm good enough to suggest pull requests because in my opinion it should be well tested. It is obvious that you are better know how RR works. Facts we have:
> 
> 1. headers that are common use don't use brackets for IPv6 to provide IP address
> 2. php, as a child of RR process expects that remote_addr should be IP address, no brackets

I'm thinking about passing the value w/o touching to the PHP worker. Since `proxy_ip_parser` parses it (even w/o sq. brackets), but `http` plugin rejecting that value (`net.SplitHostPort` expects square brackets in the IP).

**Reply by rustatian** [reply]
```go
func FetchIP(pair string) string {
	if !strings.ContainsRune(pair, ':') {
		return pair
	}
	addr, _, err := net.SplitHostPort(pair)
	if err == nil {
		return addr
	}
	ip := net.ParseIP(pair)
	if ip == nil {
		return ""  // or other error processing in such case
	}
	return ip.String()
}
```

This code snippet I guess also may fix the issue, but all methods used in this function are very slow and I'm not sure yet, what is the better option:
1. Pass as is (direct value from the `proxy_ip_parser`).
2. Test and apply this code snippet.
3. Tune the `proxy_ip_parser` middleware and add the sq.brackets to the IP address from the headers.

**Reply by rustatian** [reply]
Ok guys, I'm fine with the option 2. I've discussed this change with our PHP team (since I'm not a PHP developer) and option 1 will lead to a BC and option 3 will lead to unnecessary edge cases in parsing+adding sq.brackets.

Adding additional `net.ParseIP` will slow down the execution a little, but only in case when CF/XFF headers are used.

Thanks for noting that.

**Reply by rustatian** [reply]
ref: https://github.com/roadrunner-server/http/pull/114

**Reply by vladimir-vv** [reply]
I like you want RR being fast. How about a bit simplify FetchIP()
For example before row with net.ParseIP() add these:
```golang
	if strings.Contains(pair, "::") && !strings.HasPrefix(pair, "[") {
		return pair
	}
```
Or something like that? I think you know what will be better. I didn't even think that it may slow down request processing.
Thank you 🙏

**Reply by vladimir-vv** [reply]
Thank you very much 🙏
We are really appreciate it.

**Reply by rustatian** [reply]
The function `SplitHostPort` already checks this case with the square brackets.

From an algorithmic point of view, this is basically the same: O(n) for the strings.Contains + O(1) for checking string[0] (`SplitHostPort` also checks the string[0] index) = O(n).

From a latency perspective, this is about 0.1% of PHP time. For example, IPC communication costs us 100ms, but adding `net.ParseIP` adds 1ms. If we optimize this to 0.5ms, we would still have 100.5ms instead of 101ms 😃.

I will release the final release candidate this Thursday so you can test the changes. Thanks a lot!

**Reply by vladimir-vv** [reply]
Sorry if I'm bothering you

```golang
func FetchIP(pair string) string {
	if !strings.ContainsRune(pair, ':') {
		return pair
	}
	addr, _, err := net.SplitHostPort(pair)
	if err == nil {
		return addr
	}
	// will be here in case if **pair** is just a IPv6 address without brackets
	return pair
}
```
If we suppose that **pair** is ~ golang RemoteAddr than it will be enough to return **pair** in case if SplitHostPort return error
What do you think?
;-)

**Reply by rustatian** [reply]
NP, you're not bothering me, feel free to ask questions/ping me 😃 

I think that we should additionally check the provided IP address. In case if user uses only `http` plugin w/o a middleware, `SplitHostPort` is sufficient because it on a par with the `RemoteAddr`. So, any `RemoteAddr` would be parsed succesfully with the `SplitHostPort`. On the other hand, since `proxy_ip_parser` rewrites this address (`RemoteAddr`), we need to additionally verify it, since my super XFF/CF 🥲 parser might fail haha))

Just FYI, RoadRunner is not a replacement for the web server. Since we have some of its basic functionality, I don't intend to replace `NGINX` or other web servers with RoadRunner, which is an application server.

**Reply by rustatian** [reply]
@vladimir-vv BTW2: If you wanted to test this change -> all binaries that passed the tests might be found here: https://github.com/roadrunner-server/roadrunner/actions/runs/4552603327 (in the CI artifacts)

**Reply by vladimir-vv** [reply]
Looks like its working! :+1:  
Thank you!

**Comment by victor-sudakov** [comment]
Thank you @vladimir-vv for isolating the issue and suggesting a fix, and @rustatian for implementing it! Now this RR middleware is useful behind CloudFlare with IPv6.

**Comment by victor-sudakov** [comment]
> RoadRunner is not a replacement for the web server. Since we have some of its basic functionality, I don't intend to replace NGINX or other web servers with RoadRunner, which is an application server.

Sure, RoadRunner works very well behind a Kubernetes ingress. It's ideal to serve Laravel apps.
